pvlib/modelchain.py

```diff
@@ -961,6 +1035,97 @@ def run_model(self, weather, times=None):
         self.aoi_model()
         self.spectral_model()
         self.effective_irradiance_model()
+
+        self.run_model_from_effective_irradiance()
+
+        return self
+
+    def run_model_from_poa(self, total_irrad):
+        """
+        Run the model starting with broadband irradiance in the plane of array.
+
+        Inputs must include direct, diffuse and total irradiance (W/m2) in the
+        plane of array. Reflections and spectral adjustments are made to
+        calculate effective irradiance (W/m2).
+
+        Attribute `ModelChain.weather` can include columns ``'temp_air'`` and
+        ``'wind_speed'``; if not provided, air temperature of 20 C and wind
+        speed of 0 m/s are added to `weather`.
+
+        This method calculates:
+            * effective irradiance
+            * cell temperature
+            * DC output
+            * AC output
+
+        Parameters
+        ----------
+        total_irrad : DataFrame
+            Column names must include ``'poa_global'``, ``'poa_direct'``
+            ``'poa_diffuse'``.
+
+        Returns
+        -------
+        self
+
+        Assigns attributes: ``weather``, ``effective_irradiance``,
+        ``cell_temperature``, ``dc``, ``ac``, ``losses``, ``diode_params``
+        (if dc_model is a single diode model)
+        """
+
+        self.prepare_inputs_from_poa(total_irrad)
+
+        self.aoi_model()
+        self.spectral_model()
+        self.effective_irradiance_model()
+
+        self.run_model_from_effective_irradiance()
+
+        return self
+
+    def run_model_from_effective_irradiance(self):
+        """
+        Run the model starting with effective irradiance in the plane of array.
+
+        Attribute `ModelChain.effective_irradiance` is required.
+
+        Plane-of-array irradiance is broadband irradiance (W/m2) in the plane
+        of array without accounting for soiling or reflections. Effective
+        irradiance (W/m2) takes into account soiling, reflections and spectrum
+        adjustments.
+
+        Attribute `ModelChain.weather` (DataFrame) can include columns
+        ``'wind_speed'`` and ``'temp_air'``; if not, air temperature of 20 C
+        and wind speed of 0 m/s will be added to `weather`.
+
+        This method calculates:
+            * cell temperature
+            * DC output
+            * AC output
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+        self
+
+        Assigns attributes: ``cell_temperature``, ``dc``, ``ac``,
+        ``losses``, ``diode_params`` (if dc_model is a single diode model)
+        """
+
+        # assign poa_global column for the temperature model
+        if 'poa_global' not in self.total_irrad.columns:
+            try:
+                self.total_irrad['poa_global'] = \
+                    self.effective_irradiance
+            except KeyError:
```

Comment:
I don't understand what KeyError this is supposed to catch.

Reply:
Nor do I